resource/file/file.go

```diff
@@ -0,0 +1,566 @@
+	ActionDelete
+)
+
+func (a Action) String() string {
+	return string(a)
+}
+
+// State is the desired state of the file
+type State string
+
+const (
+	// StateAbsent indicates the file should be absent
+	StateAbsent State = "absent"
+
+	// StateUndefined indicates an undefined state
+	StateUndefined State = ""
+
+	// StatePresent indicates the file should be present
+	StatePresent State = "present"
+
+	// DefaultState is the default state
+	DefaultState State = StatePresent
+)
+
+// ValidStates indicates supported states that can be defined
+var ValidStates = []State{StateAbsent, StatePresent}
+
+func (s State) String() string {
+	return string(s)
+}
+
+// Type is the type of file
+type Type string
+
+const (
+	// TypeDirectory is a directory
+	TypeDirectory Type = "directory"
+
+	// TypeFile is a regular file
+	TypeFile Type = "file"
+
+	// TypeLink is a hardlink
+	TypeLink Type = "hardlink"
+
+	// TypeNone is an undefined type
+	TypeNone Type = ""
+
+	// TypeSymlink is a symlink
+	TypeSymlink Type = "symlink"
+
+	// DefaultType is the default type
+	DefaultType Type = TypeFile
+)
+
+var (
+	// ValidFileTypes are valid file types
+	ValidFileTypes = []Type{TypeDirectory, TypeFile}
+
+	// ValidLinkTypes are valid link types
+	ValidLinkTypes = []Type{TypeLink, TypeSymlink}
+
+	// AllTypes are the current FileTypes that are supported
+	AllTypes = append(ValidFileTypes, ValidLinkTypes...)
+)
+
+func (t Type) String() string {
+	return string(t)
+}
+
+// File contains information for managing files
+type File struct {
+	Destination string
+	State       State
+	Type        Type
+	Target      string
+	Force       bool    //force replacement of file
+	Mode        *uint32 //requested permissions from Prepare
+	UserInfo    *user.User
+	GroupInfo   *user.Group
+	Content     []byte
+
+	action Action //create, delete, modify
+
+	renderer resource.Renderer
+}
+
+// New returns a File with UserInfo and GroupInfo fields allocated
+func New() *File {
+	return &File{
+		UserInfo:  &user.User{},
+		GroupInfo: &user.Group{},
+	}
+}
+
+// Apply changes to file resources
+func (f *File) Apply() (resource.TaskStatus, error) {
+
+	status, err := f.diff()
+	if err != nil || status.Level == resource.StatusFatal {
+		return status, errors.Wrap(err, "apply")
+	}
+
+	switch f.action {
+	case ActionDelete:
+		err = f.Delete()
+	case ActionCreate, ActionModify:
+		err = f.Modify(status)
+	}
+	if err != nil {
+		status.Level = resource.StatusFatal
+		status.AddMessage(err.Error())
+	}
+	return status, err
+}
+
+// Check File settings
+func (f *File) Check(r resource.Renderer) (resource.TaskStatus, error) {
+
+	f.renderer = r
+
+	status, err := f.diff()
+	status.AddMessage(fmt.Sprintf("%s (%s)", f.Destination, f.Type))
+	if err != nil || status.Level == resource.StatusFatal {
+		return status, errors.Wrap(err, "check")
+	}
+
+	return status, err
+}
+
+// Compare desired state with an actual file (if present)
+func (f *File) diff() (*resource.Status, error) {
```

Comment:
There is a lot of deeply nested logic here with a number of exit points.  Is there any way to factor some of the logic out into functions so that `diff()` itself is just managing the branching logic and the functions are managing the other logic?